tests/MongoDB.Driver.Tests/PinnedShardRouterTests.cs

```diff
@@ -172,11 +172,17 @@ private DisposableMongoClient CreateDisposableClient(EventCapturer eventCapturer
             var client = DriverTestConfiguration.CreateDisposableClient((MongoClientSettings settings) =>
                 {
                     settings.ClusterConfigurator = c => c.Subscribe(eventCapturer);
-                    settings.LocalThreshold = TimeSpan.FromMilliseconds(1000);
+                    settings.LocalThreshold = TimeSpan.FromMilliseconds(5000);
                 }, 
                 useMultipleShardRouters);
             var timeOut = TimeSpan.FromSeconds(60);
-            SpinWait.SpinUntil(() => client.Cluster.Description.Type != ClusterType.Unknown, timeOut).Should().BeTrue();
+
+            var cluster = client.Cluster;
+            SpinWait.SpinUntil(condition: () =>
+                cluster.Description.Type != ClusterType.Unknown
+                && cluster.Description.Servers.Count >= (useMultipleShardRouters ? 2 : 1)
+                && !(cluster.Description.Servers.Select(s => s.State != ServerState.Connected).Any()),
+                timeOut).Should().BeTrue();
```

Comment:
I think the following one liner might be sufficient:

    SpinWait.SpinUntil(() => client.Cluster.Description.Servers.All(s => s.State == ServerState.Connected), timeOut).Should().BeTrue();

Reply:
I've taken your change with a very minor formatting tweak :)